### PR TITLE
Feature - Line highlighting within code block #2469

### PR DESCRIPTION
**Pull request by duartefrazao**
## Description
This PR solves the #2469 issue, it allows the user to highlight lines of code within a snippet of code by clicking on the line number. This highlight is not temporary, it persists when exiting Boostnote. Used CodeMirror built-in functionality to do this.

To use this feature you just need to click the line number of the respective line you want to highlight.

### Before
screenshot from 2018-12-05 13-19-51
### After
screenshot from 2018-12-05 13-20-18

## Issue fixed
Issue #2469 Line highlighting within code block

## Type of changes

- :radio_button: Bug fix (Change that fixed an issue)
- :white_circle: : Breaking change (Change that can cause existing functionality to change)
- :radio_button: Improvement (Change that improves the code. Maybe performance or development improvement)
- :radio_button: Feature (Change that adds new functionality)
- :white_circle: Documentation change (Change that modifies documentation. Maybe typo fixes)

## Checklist:

- :radio_button: My code follows [the project code style](docs/code_style.md)
- :white_circle: I have written test for my code and it has been tested
- :radio_button: All existing tests have been passed
- :radio_button: I have attached a screenshot/video to visualize my change if possible

### CODE REVIEW

File: browser/components/CodeEditor.js [hunk 1]
```diff
@@ -683,6 +708,24 @@ export default class CodeEditor extends React.Component {
     })
   }
 
+  initialHighlighting () {
+    if (this.editor.options.linesHighlighted == null) {
+      return
+    }
+
+    const count = this.editor.lineCount()
+    for (let i = 0; i < count; i++) {
```
Comment (ZeroX-DG) [inline browser/components/CodeEditor.js]: Will it be faster if we just loop the highlighted lines?
```js
const totalHighlightedLines = this.editor.options.linesHighlighted.length;
const totalAvailableLines = this.editor.lineCount();
for (let i = 0; i < totalHighlightedLines; i++){
  const lineNumber = this.editor.options.linesHighlighted[i];
  if (lineNumber > totalAvailableLines) {
    // make sure that we skip the invalid lines althrough this case should not be happened.
   continue;
  }
  this.editor.addLineClass(lineNumber, 'text', 'CodeMirror-activeline-background');
}
```

Comment (duartefrazao) [reply browser/components/CodeEditor.js]: Thank you for the fast response and review.
Yes, your changes make it better optimized.
I will make the changes to the code.

**Comment by ZeroX-DG** [comment]
Hmm...That's strange, I can't highlight the line by clicking on the gutter. Can you check it?

**Comment by ZeroX-DG** [comment]
Please fix your eslint error

**Comment by ZeroX-DG** [comment]
Also I discover that when my snippet doesn't contain `linesHighlighted` then the feature won't work at all. You should fallback `linesHighlighted` to `[]` if it not existed. Also, can you apply this feature to markdown note too?

**Comment by duartefrazao** [comment]
> Please fix your eslint error

Sorry, fixed.

**Comment by ZeroX-DG** [comment]
I can still reproduce the old bug. You should fallback `linesHighlighted` to `[]` if it is null instead of stopping the whole process.

**Comment by duartefrazao** [comment]
Could you please tell me how to reproduce that bug?
Is it a default note, an old note pre pull request or a note that you changed?

**Comment by ZeroX-DG** [comment]
It's an old default snippet note generated by Boostnote.

**Comment by duartefrazao** [comment]
I've been trying to replicate this old bug but with no sucess. 
I've tried to delete everyting Boostone related and download a past version and then open with my version of the pull request and it still works...
Could you detail the steps you took as well as operating system you're using?

**Comment by duartefrazao** [comment]
I've tested in several computers and I can't replicate the bug anymore, please give me detailed instructions to replicate your bug.

**Comment by ZeroX-DG** [comment]
I have to manually added `linesHighlighted` to the first note and leave the second note alone (the default note of Boostnote). 
bbb

**Comment by duartefrazao** [comment]
This new commit contains two main parts:
- Included the data structure holding past values of the lines highlighted for each note/snippet saved in the create and update snippet/note tests 
- Made the feature robust to the following variables when editing the note/snippet:
  - Type of action (delete, input, paste, undo and insert)
  - How many lines affected (e.g. be it a single line delete or multi-line)
  - Where the action starts and ends (e.g. if you end a delete on the middle of a line the highlight should be kept but if you end on the begging of the line the highlight should be removed)

**Comment by duartefrazao** [comment]
@ZeroX-DG I'm also part of the same class that has recently contributed to Boostnote. As some of them mentioned It would really help my group's grade if we knew if the pull request is going to be accepted. Thank you very much!

File: browser/components/MarkdownEditor.js [hunk 2]
```diff
@@ -275,7 +275,8 @@ class MarkdownEditor extends React.Component {
           noteKey={noteKey}
           fetchUrlTitle={config.editor.fetchUrlTitle}
           enableTableEditor={config.editor.enableTableEditor}
-          onChange={(e) => this.handleChange(e)}
+          linesHighlighted={linesHighlighted}
+          onChange={(e) => this.handleChange.bind(this)(e)}
```
Comment (ZeroX-DG) [inline browser/components/MarkdownEditor.js]: You don't need to `bind(this)` in an arrow function, the `this` keyword will point to the class if you use arrow function.

File: browser/components/MarkdownSplitEditor.js [hunk 3]
```diff
@@ -169,7 +169,8 @@ class MarkdownSplitEditor extends React.Component {
           enableTableEditor={config.editor.enableTableEditor}
           storageKey={storageKey}
           noteKey={noteKey}
-          onChange={this.handleOnChange.bind(this)}
+          linesHighlighted={linesHighlighted}
+          onChange={(e) => this.handleOnChange.bind(this)(e)}
```
Comment (ZeroX-DG) [inline browser/components/MarkdownSplitEditor.js]: Right here too, please change it to `{(e) => this.handleOnChange(e)}`

File: browser/components/MarkdownEditor.js (hunk 2 above)
Comment (duartefrazao) [reply browser/components/MarkdownEditor.js]: Didn't know that, thank you for the correction!

**Comment by duartefrazao** [comment]
Made the changes you asked, @ZeroX-DG . Should have asked when that appeared on the Markdown side, thank you for explaining :smiley:

**Comment by duartefrazao** [comment]
Thank you very much, and thank you for guiding us on submitting our pull request! Hope our work benefits Boostnote :smiley:

**Pull request merged**